8252888: Collapse G1MMUTracker class hierarchy

JohnTortugo opened the pull request:
Relates to: https://bugs.openjdk.java.net/browse/JDK-8252888
Tested on: x86_64 - Linux - tier1
Need sponsor: yes

Merge `G1MMUTracker` and `G1MMUTrackerQueue` into a single class and make method members non virtual.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252888](https://bugs.openjdk.java.net/browse/JDK-8252888): Collapse G1MMUTracker class hierarchy


### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 136a1ae78ac6e3b4abe8f6e595431b5c0696566b
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/214/head:pull/214`
`$ git checkout pull/214`

albertnetymk commented:
I think it's better to use `G1MMUTracker` as the type name and mention its queue implementation in the comments, instead of including "Queue" in its name.

JohnTortugo commented:
@albertnetymk - Thanks for reviewing. I just renamed accordingly to your suggestion. 

I also renamed MMUTrackerQueueElement to MMUTrackerElement as it made sense to me. Please let me know if you disagree and I can revert that.

JohnTortugo commented:
/test

albertnetymk reviewed (approved):
Thank you for the update. If you feel like it, `static consexpr uint QueueLength = 64;` reads better than `enum PrivateConstants { ... }`, IMO. Not sure why an `enum` was used in the first place. The rest look fine to me.

albertnetymk commented:
> https://bugs.openjdk.java.net/browse/JDK-8153116

Kim, thank you for the info.

kstefanj commented on src/hotspot/share/gc/g1/g1MMUTracker.hpp:
```diff
@@ -77,25 +39,43 @@ class G1MMUTrackerQueueElem {
-// that keeps track of all the recent pause times
-class G1MMUTrackerQueue: public G1MMUTracker {
+
+// This implementation uses a fixed-size queue to keep track of all
+// recent pause times. Two major user controls over G1 behavior are
```
I would move the first sentence down to right before the warning about seconds. I would also change it to something like:
`The G1MMUTracker uses a fixed-size queue to keep track of all recent pause times. The pause time data is used to avoid breaking the MMU.`

JohnTortugo replied:
done

kstefanj reviewed (changes requested):
Thanks for doing this cleanup. 

Looks good, please just do a small change to one of the comments.

kstefanj commented:
I'm currently running some basic testing to make sure everything is fine.

kstefanj commented:
Testing looks good.

JohnTortugo commented:
@kstefanj - I addressed your feedback. Thanks for reviewing. PTAL when you have some time.

kstefanj commented:
> @kstefanj - I addressed your feedback. Thanks for reviewing. PTAL when you have some time.

I see there are some whitespace issues, once those are fixed I'm ok with this fix. But as Kim I wouldn't mind if the the extra indenting of the members was removed instead of pushed right even more.

JohnTortugo commented:
/integrate

kstefanj reviewed (approved):
Thanks for addressing our comments, looks good.

kstefanj commented:
/sponsor
